src/main/kotlin/com/varabyte/kobweb/intellij/colors/KobwebColorProvider.kt

```diff
@@ -0,0 +1,222 @@
+@file:Suppress("UseJBColor")
+
+package com.varabyte.kobweb.intellij.colors
+
+import com.intellij.ide.util.PsiNavigationSupport
+import com.intellij.openapi.editor.ElementColorProvider
+import com.intellij.openapi.fileEditor.OpenFileDescriptor
+import com.intellij.psi.PsiElement
+import com.intellij.psi.impl.source.tree.LeafPsiElement
+import com.intellij.psi.util.PsiUtilBase
+import org.jetbrains.kotlin.analysis.api.analyze
+import org.jetbrains.kotlin.analysis.api.components.KtConstantEvaluationMode
+import org.jetbrains.kotlin.idea.base.psi.kotlinFqName
+import org.jetbrains.kotlin.idea.core.util.toPsiFile
+import org.jetbrains.kotlin.idea.references.mainReference
+import org.jetbrains.kotlin.js.translate.declaration.hasCustomGetter
+import org.jetbrains.kotlin.lexer.KtTokens
+import org.jetbrains.kotlin.psi.*
+import java.awt.Color
+import kotlin.math.abs
+
+// This prevents the color tracer from checking the codebase ridiculously deep,
+// which might cause lag or even worse a stackoverflow
+private const val MAX_SEARCH_DEPTH = 15
+
+private const val KOBWEB_COLOR_COMPANION_FQ_NAME = "com.varabyte.kobweb.compose.ui.graphics.Color.Companion"
+
+/**
+ * Enables showing those small rectangular color previews next to the line number
+ */
+class KobwebColorProvider : ElementColorProvider {
+
+    override fun getColorFrom(element: PsiElement): Color? = when {
+        element !is LeafPsiElement -> null
+        element.elementType != KtTokens.IDENTIFIER -> null
+        element.parent is KtProperty -> null // Avoid showing multiple previews
+        else -> traceColor(element.parent) // Leaf is just text. The parent is the actual object
+    }
+
+    /**
+     * Setting colors is not (yet) supported
+     */
+    override fun setColorTo(element: PsiElement, color: Color) = Unit
+}
+
+/**
+ * Tries resolving references as deep as possible and checks if a Kobweb color is being referred to
+ */
+private fun traceColor(element: PsiElement, currentDepth: Int = 0): Color? {
+    val nextElement = when (element) {
+        is KtDotQualifiedExpression -> element.selectorExpression
+
+        is KtNameReferenceExpression -> when {
+            element.parent is KtCallExpression -> element.parent // Element is name of a function
+            else -> element.locateSource()
+        }
+
+        is KtProperty -> when {
+            element.hasInitializer() -> element.initializer
+            element.hasCustomGetter() -> element.getter
+            else -> null
+        }
+
+        is KtPropertyAccessor -> element.bodyExpression
+
+        is KtCallExpression -> (null).apply {
```

Comment:
`(null).apply` feels pretty weird to me. I've never seen this before once, so I'm just going to double check here. My intuition says there's probably a clearer way to express what you're doing; I suppose it's possible it's some cool new pattern that I just need to be exposed to to understand its brilliance!

What are you trying to do here? I think what you want to do is put a `run { ... }`?

Reply:
Yeah I don't think there is any reason behind this cause I wrote this very late and night and then just didn't notice it again lol

Reply:
I just figured out again what this was supposed to to. I wanted to return null in any case but still wanted to run the color method detection code.

Reply:
Actually thinking about it makes me like my approach more, however I would rename `apply` to `also`.

```kt
is KtCallExpression -> null.also { ... }
```

so we **return null** for this when branch and **also** run the color function detection code.

What is your thought on this?